src/track/serato/markers2.cpp

```diff
@@ -391,5 +393,69 @@ QByteArray SeratoMarkers2::data() const {
     return outerData.leftJustified(size, '\0');
 }
 
+void SeratoMarkers2::syncToTrackObject(Track* track) const {
+    qDebug() << "Syncing 'Serato Markers2' tag data to Track object...";
+
+    QMap<int, const mixxx::SeratoMarkers2CueEntry*> cueMap;
+    for (auto& pEntry : m_entries) {
+        DEBUG_ASSERT(pEntry);
+        switch (pEntry->typeId()) {
+        case SeratoMarkers2Entry::TypeId::Cue: {
+            const mixxx::SeratoMarkers2CueEntry* pCueEntry = static_cast<mixxx::SeratoMarkers2CueEntry*>(pEntry.get());
+            cueMap.insert(pCueEntry->getIndex(), pCueEntry);
+            break;
+        }
+        case SeratoMarkers2Entry::TypeId::Bpmlock: {
+            const mixxx::SeratoMarkers2BpmlockEntry* pBpmlockEntry = static_cast<mixxx::SeratoMarkers2BpmlockEntry*>(pEntry.get());
+            track->setBpmLocked(pBpmlockEntry->isLocked());
+            break;
+        }
+        // TODO: Add support for COLOR/LOOP/FLIP
+        default:
+            break;
+        }
+    }
+
+    const double sampleRateKhz = track->getSampleRate() / 1000.0;
+    const double samples = sampleRateKhz * mixxx::kEngineChannelCount;
+    VERIFY_OR_DEBUG_ASSERT(samples != 0) {
+        qWarning() << "Failed to calculate samples factor, cannot import cues!";
+        return;
+    }
+
+    // Update existing hotcues
+    for (const CuePointer& pCue : track->getCuePoints()) {
```

Comment:
One option would be to introduce a separate, independent "SeratoMarkers2TrackSync" class that is responsible for performing the synchronization. It is allowed to have dependencies to both `Track` and `SeratoMarkers2`. This would avoid the cursed cyclic dependencies.

The whole `Track` object design is broken and a big ball of mud. Try to isolate synchronization logic as much as possible. Don't add new dependencies on the omnipresent `Track` UI object to innocent, new classes ;)

Reply:
That's what I'm planning to do with the SeratoTags class. Since the tags contain redundant and possibly conflicting information, that might be a good idea anyways.